### ISSUE TITLE
nil dereference when created timestamp is missing in image history

### ISSUE THREAD

**Issue by wolf-cosmose** [moby/moby#47729]
### Description

I encountered a docker image for which displaying history wouldn't work:

```console
$ docker history REDACTED
error during connect: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.44/images/REDACTED/history": EOF
```

I looked into dockerd logs and found the following error:
```
http: panic serving @: runtime error: invalid memory address or nil pointer dereference
```
With a stacktrace:
<details>

```
2024/04/18 17:19:52 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
goroutine 2717 [running]:
net/http.(*conn).serve.func1()
        net/http/server.go:1868 +0xb9
panic({0x6113b1b631a0?, 0x6113b33645a0?})
        runtime/panic.go:920 +0x270
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
        go.opentelemetry.io/otel/sdk@v1.19.0/trace/span.go:383 +0x25
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc000f99380, {0x0, 0x0, 0x6113b1fdd220?})
        go.opentelemetry.io/otel/sdk@v1.19.0/trace/span.go:421 +0x9fb
panic({0x6113b1b631a0?, 0x6113b33645a0?})
        runtime/panic.go:914 +0x21f
github.com/docker/docker/daemon/images.(*ImageService).ImageHistory(0xc000fa4dc0, {0x6113b1fdb000, 0xc000e76ae0}, {0xc000fbc012, 0x40})
        github.com/docker/docker/daemon/images/image_history.go:47 +0x423
github.com/docker/docker/api/server/router/image.(*imageRouter).getImagesHistory(0xc000037a40, {0x6113b1fdb000, 0xc000e76ae0}, {0x6113b1fd18e0, 0xc0026e4120}, 0x6113b1484cc4?, 0xc000e763f0?)
        github.com/docker/docker/api/server/router/image/image_routes.go:432 +0x72
github.com/docker/docker/api/server/middleware.(*ExperimentalMiddleware).WrapHandler.ExperimentalMiddleware.WrapHandler.func1({0x6113b1fdb000, 0xc000e76ae0}, {0x6113b1fd18e0?, 0xc0026e4120?}, 0x6113b1a080e0?, 0xc001dc4100?)
        github.com/docker/docker/api/server/middleware/experimental.go:26 +0xb4
github.com/docker/docker/api/server/middleware.(*VersionMiddleware).WrapHandler.VersionMiddleware.WrapHandler.func1({0x6113b1fdb000, 0xc000e76ab0}, {0x6113b1fd18e0, 0xc0026e4120}, 0x40?, 0x40?)
        github.com/docker/docker/api/server/middleware/version.go:62 +0x2ae
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1({0x6113b1fdb000, 0xc000e76ab0}, {0x6113b1fd18e0?, 0xc0026e4120?}, 0xc001fa2600, 0xc001dc4070?)
        github.com/docker/docker/pkg/authorization/middleware.go:59 +0x683
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1({0x6113b1fd18e0, 0xc0026e4120}, 0xc001fa2500)
        github.com/docker/docker/api/server/server.go:55 +0x1c3
net/http.HandlerFunc.ServeHTTP(0x6113b1fdb000?, {0x6113b1fd18e0?, 0xc0026e4120?}, 0x6113b183e078?)
        net/http/server.go:2136 +0x29
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP(0xc0013faa50, {0x6113b1fc9630?, 0xc000db02a0}, 0xc001fa2200, {0x6113b1fb02a0, 0xc00019f458})
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.45.0/handler.go:217 +0x1202
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1({0x6113b1fc9630?, 0xc000db02a0?}, 0x6113b1837101?)
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.45.0/handler.go:81 +0x35
net/http.HandlerFunc.ServeHTTP(0x6113b1fdb000?, {0x6113b1fc9630?, 0xc000db02a0?}, 0x6113b18371a0?)
        net/http/server.go:2136 +0x29
net/http.HandlerFunc.ServeHTTP(0xc001fa2100?, {0x6113b1fc9630?, 0xc000db02a0?}, 0x7ae39b7d2250?)
        net/http/server.go:2136 +0x29
github.com/gorilla/mux.(*Router).ServeHTTP(0xc000b400c0, {0x6113b1fc9630, 0xc000db02a0}, 0xc001fa2000)
        github.com/gorilla/mux@v1.8.1/mux.go:212 +0x1c5
net/http.serverHandler.ServeHTTP({0xc000f0a570?}, {0x6113b1fc9630?, 0xc000db02a0?}, 0x6?)
        net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc0012461b0, {0x6113b1fdb000, 0xc000e141e0})
        net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 182
        net/http/server.go:3086 +0x5cb
2024/04/18 17:19:52 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
goroutine 2688 [running]:
net/http.(*conn).serve.func1()
        net/http/server.go:1868 +0xb9
panic({0x6113b1b631a0?, 0x6113b33645a0?})
        runtime/panic.go:920 +0x270
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
        go.opentelemetry.io/otel/sdk@v1.19.0/trace/span.go:383 +0x25
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc00168f080, {0x0, 0x0, 0x6113b1fdd220?})
        go.opentelemetry.io/otel/sdk@v1.19.0/trace/span.go:421 +0x9fb
panic({0x6113b1b631a0?, 0x6113b33645a0?})
        runtime/panic.go:914 +0x21f
github.com/docker/docker/daemon/images.(*ImageService).ImageHistory(0xc000fa4dc0, {0x6113b1fdb000, 0xc000e76e70}, {0xc000fbc0f2, 0x40})
        github.com/docker/docker/daemon/images/image_history.go:47 +0x423
github.com/docker/docker/api/server/router/image.(*imageRouter).getImagesHistory(0xc000037a40, {0x6113b1fdb000, 0xc000e76e70}, {0x6113b1fd18e0, 0xc0026e4300}, 0x6113b1484cc4?, 0xc000e76c90?)
        github.com/docker/docker/api/server/router/image/image_routes.go:432 +0x72
github.com/docker/docker/api/server/middleware.(*ExperimentalMiddleware).WrapHandler.ExperimentalMiddleware.WrapHandler.func1({0x6113b1fdb000, 0xc000e76e70}, {0x6113b1fd18e0?, 0xc0026e4300?}, 0x6113b1a080e0?, 0xc001dc45e0?)
        github.com/docker/docker/api/server/middleware/experimental.go:26 +0xb4
github.com/docker/docker/api/server/middleware.(*VersionMiddleware).WrapHandler.VersionMiddleware.WrapHandler.func1({0x6113b1fdb000, 0xc000e76e40}, {0x6113b1fd18e0, 0xc0026e4300}, 0x40?, 0x40?)
        github.com/docker/docker/api/server/middleware/version.go:62 +0x2ae
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1({0x6113b1fdb000, 0xc000e76e40}, {0x6113b1fd18e0?, 0xc0026e4300?}, 0xc001fa2c00, 0xc001dc43c0?)
        github.com/docker/docker/pkg/authorization/middleware.go:59 +0x683
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1({0x6113b1fd18e0, 0xc0026e4300}, 0xc001fa2b00)
        github.com/docker/docker/api/server/server.go:55 +0x1c3
net/http.HandlerFunc.ServeHTTP(0x6113b1fdb000?, {0x6113b1fd18e0?, 0xc0026e4300?}, 0x6113b183e078?)
        net/http/server.go:2136 +0x29
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP(0xc0013faa50, {0x6113b1fc9630?, 0xc000db0460}, 0xc001fa2900, {0x6113b1fb02a0, 0xc00019f458})
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.45.0/handler.go:217 +0x1202
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1({0x6113b1fc9630?, 0xc000db0460?}, 0x6113b1837101?)
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.45.0/handler.go:81 +0x35
net/http.HandlerFunc.ServeHTTP(0x6113b1fdb000?, {0x6113b1fc9630?, 0xc000db0460?}, 0x6113b18371a0?)
        net/http/server.go:2136 +0x29
net/http.HandlerFunc.ServeHTTP(0xc001fa2800?, {0x6113b1fc9630?, 0xc000db0460?}, 0x7ae39b7d2250?)
        net/http/server.go:2136 +0x29
github.com/gorilla/mux.(*Router).ServeHTTP(0xc000b400c0, {0x6113b1fc9630, 0xc000db0460}, 0xc001fa2700)
        github.com/gorilla/mux@v1.8.1/mux.go:212 +0x1c5
net/http.serverHandler.ServeHTTP({0xc000e76c00?}, {0x6113b1fc9630?, 0xc000db0460?}, 0x6?)
        net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc001188120, {0x6113b1fdb000, 0xc000e141e0})
        net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 182
        net/http/server.go:3086 +0x5cb

```

</details>

It seems like the panic occurs in this line:
https://github.com/moby/moby/blob/fce6e0ca9bc000888de3daa157af14fa41fcd0ff/daemon/images/image_history.go#L47

After exporting the image with `docker image save and inspecting the config blob, I noticed the last two history entries (after formatting with jq) look like this:
```json
    {
      "created_by": "WORKDIR /app",
      "comment": "buildkit.dockerfile.v0",
      "empty_layer": true
    },
    {
      "created": "2021-11-02T17:05:56.101418786Z",
      "created_by": "COPY . . # buildkit",
      "comment": "buildkit.dockerfile.v0"
    }
```

The empty layer created by buildkit seems to be missing a `created` timestamp. There are more layers like this in that image.
Other images build from different versions of the same project have `created` timestamps on empty layers too, so I'm guessing they're supposed to be present, and are missing on this image due to some rare bug.

I could try to figure out how such a faulty image was built, but it'll take more work and I'm not sure I'll be able to reproduce it.
The image was built 3 years ago on our CI system, then pushed to AWS ECR registry, and then fetched again.
We only started having issues with it after we updated docker on our CI system from 20 to 26 (the issue happens on both 26 and 25, but it was more convenient for me to reproduce on a machine which has docker 25).

Regardless of how such an image came into being, I think the way docker panics on it is still a bug.

I made a minimal image that reproduces the issue: [missing-timestamp-repro.tgz](https://github.com/moby/moby/files/15027216/missing-timestamp-repro.tgz)

it's made from the following Dockerfile:
```Dockerfile
FROM scratch

ADD test.txt /a/test.txt

WORKDIR /a
```

after which its config has been modified to look like this:

```json
{
  "architecture": "amd64",
  "config": {
    "Env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    ],
    "WorkingDir": "/a",
    "OnBuild": null
  },
  "created": "2024-04-18T16:48:51.558393517+02:00",
  "history": [
    {
      "created": "2024-04-18T16:48:51.540379184+02:00",
      "created_by": "ADD test.txt /a/test.txt # buildkit",
      "comment": "buildkit.dockerfile.v0"
    },
    {
      "created_by": "WORKDIR /a",
      "comment": "buildkit.dockerfile.v0"
    }
  ],
  "os": "linux",
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:12290672965f54c8d3610022e389b915a70b508ad262daf91cd178df376dfddd",
      "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef"
    ]
  }
}
```

### Reproduce

download the attached [missing-timestamp-repro.tgz](https://github.com/moby/moby/files/15027216/missing-timestamp-repro.tgz)
then run the following commands:
```console
$ tar -xzvf  missing-timestamp-repro.tgz
missing-timestamp-repro/
missing-timestamp-repro/missing-timestamp.docker.tar
$ docker image load -i missing-timestamp-repro/missing-timestamp.docker..tar
Loaded image ID: sha256:a6fb7d75307640128bf713822664aa0cf92c3007acdeea46d85418d628b2675e
$ docker history sha256:a6fb7d75307640128bf713822664aa0cf92c3007acdeea46d85418d628b2675e
error during connect: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.44/images/sha256:a6fb7d75307640128bf713822664aa0cf92c3007acdeea46d85418d628b2675e/history": EOF
$ journalctl -u docker  -e
[...]
http: panic serving @: runtime error: invalid memory address or nil pointer dereference
[...]
```

### Expected behavior

Either the `docker image load` should report an error and refuse to load the image,
or `docker history` should display a history with missing / placeholder timestamp for the layer with missing `created`.

### docker version

```bash
Client:
 Version:           25.0.2
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629222
 Built:             Thu Feb  1 10:50:44 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          25.0.2
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       fce6e0ca9b
  Built:            Thu Feb  1 10:50:44 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.13
  GitCommit:        7c3aca7a610df76212171d200ca3811ff6096eb8.m
 runc:
  Version:          1.1.12
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Version:    25.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  0.12.1
    Path:     /usr/lib/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  2.24.5
    Path:     /usr/lib/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 152
 Server Version: 25.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7c3aca7a610df76212171d200ca3811ff6096eb8.m
 runc version: 
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.7.4-arch1-1
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 14.86GiB
 Name: alucard
 ID: JR7S:R5ML:YC2F:CLIH:IPWQ:Z3O7:YYIC:25IG:52M7:I72S:QD6L:P645
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
So, the panic looks to be happening here; https://github.com/moby/moby/blob/v25.0.2/daemon/images/image_history.go#L47 / https://github.com/moby/moby/blob/fce6e0ca9bc000888de3daa157af14fa41fcd0ff/daemon/images/image_history.go#L47


Interestingly, that code didn't change for many years; also looking at history for that field, and I see it was changed to a pointer in the OCI spec .. 7 Years ago; https://github.com/opencontainers/image-spec/commit/96896c365cca5f9b95f36abf25bb5c6c0ede5821 surprised that we didn't run into this before!

**Comment by thaJeztah** [comment]
Oh! I think I see why; while that code hasn't changed, I now recall we did change types from our own types to the OCI ones, and that's where a difference was; the previous types did _not_ have a pointer there, and I think we missed a proper check for this case when it was changed; https://github.com/moby/moby/commit/2a6ff3c24fd790e5d42d2eabaf6acf06edfe6975

- https://github.com/moby/moby/pull/45504

That change was first released in v25.0.0-beta.1


cc @tianon in case you want to have a look at a fix (otherwise someone else can probably work on one)

**Comment by krissetto** [comment]
With the fix in [this pr](https://github.com/moby/moby/pull/47734), following the reproduction steps no longer causes issues
